### DISCUSSION TITLE
Using locate_cell() mesh method on vertex of cell

### DISCUSSION THREAD

**Discussion by jmm1828** [firedrakeproject/firedrake#3015] [Firedrake support] [answered]
I am trying to use `locate_cell` on the nodes of a mesh to see which cells they correspond to. But I ran into the problem that the method only returns one cell number. 

Does anyone now a way around this? Any help would be much appreciated.

**Comment by wence-** [comment]
If you want the topological relationship, you are better off querying the topological mesh, rather than doing geometric searches. What do you want to do with the resulting information? One way to construct this in Python is to take a piecewise linear function space `V = FunctionSpace(mesh, "P", 1)` and then manually produce the inverse of the `V.cell_node_map()` whose values you can obtain with `V.cell_node_map().values`

**Reply by jmm1828** [reply]
Thank you for this, I will try it this way. I am trying to get the nodes of each cell in order to compute a certain matrix for each cell in the mesh.

**Reply by wence-** [reply]
What is the matrix you would like to compute? Most of the time if it is finite element related it should be possible to compute it without inspecting things at this low level (as @dham notes).

**Reply by jmm1828** [reply]
For stabilisation purposes, I want to calculate unit tangent vectors along the edges of each cell, then calculate the outer product of each tangent with itself and finally add the matrices for each tangent together.

**Reply by wence-** [reply]
it seems like nearly all the machinery for this exists in tsfc so let me try later to see if it can be done at a high level

**Reply by jmm1828** [reply]
Great, thank you for your help. I will look into tsfc.

**Reply by jmm1828** [reply]
If we create a mesh, say `mesh=UnitSquareMesh(2,2)`.

Then we create the function space `V = FunctionSpace(mesh, "P", 1)` and manually produce the inverse of the `V.cell_node_map()`.

Do the values in `V.cell_node_map().values` correspond with the position of the nodes in the array `mesh.coordinates.dat.data`?

**Reply by wence-** [reply]
yes. So if the first row has entries [0, 2, 3] that means the three vertices of the first cell are found at locations 0, 2, and 3 of the coordinate data

**Reply by wence-** [reply, marked as answer]
I thought this worked. And I think it does. But the relevant UFL geometry object is not correctly exported:

```python
from firedrake import *
from ufl.classes import CellEdgeVectors # this object is not exported by UFL

mesh = UnitSquareMesh(2, 2)
cev = CellEdgeVectors(mesh)

expr = sum(outer(cev[i, :], cev[i, :]) for i in range(3))

cell_matrices = interpolate(expr, TensorFunctionSpace(mesh, "DG", 0, shape=(2, 2)))
```

Gives you what you want I think.

**Reply by jmm1828** [reply]
Thank you, it does seem to work. Is there any reason for using discontinuous Lagrange instead of any other space for `TensorFunctionSpace`?

**Reply by wence-** [reply]
Well, you want one value per cell, no?

**Reply by jmm1828** [reply]
Yes, thank you

**Comment by dham** [comment]
The purpose of `locate_cell` is to return *a* cell containing the point in question. This is to enable interpolation to work. Can you please describe more generally what you are trying to achieve. Usually, querying the internals of Firedrake is not the best way to achieve things.
